docs/language-server/streaming-file-transfer.md

```diff
@@ -0,0 +1,72 @@
+---
+layout: developer-doc
+title: Streaming File Transfer
+category: language-server
+tags: [language-server, protocol, specification]
+order: 5
+---
+
+# Streaming File Transfer
+
+Particularly important in the separate-server design that we provide with the
+Language Server is the ability to transfer files to and from the remote machine.
+To that end, it is important that we provide the ability for the IDE to both
+upload and download very large files.
+
+A few key requirements:
+
+- We want to support resumption of transfers.
+- We want to have transfers be as low-overhead as possible.
+- Multiple transfers may be occurring at once.
+- We want to keep implementation simple to avoid errors (ideally no state).
+
+<!-- MarkdownTOC levels="2,3" autolink="true" indent="    " -->
+
+- [Control](#control)
+- [UX](#ux)
+
+<!-- /MarkdownTOC -->
+
+## Control
+
+In order to make this portion of the protocol simple to manage, it is defined in
+a stateless fashion. The Language Server provides two messages `file/writeBytes`
+and `file/readByteRange` with corresponding responses. We make the following
+assumptions:
+
+- Each request must be completed with a response before sending another request
+  for the same file.
```

Comment:
While I understand that this should likely greatly simplify the protocol, isn't it an important limitation? Especially on a high-bandwidth, but also high-latency connection, a significant portion of time will be spent waiting for the next request, while it could have been transferring next chunks.

Or are the chunks that are requested meant to be so large that this issue is insignificant?